cli/src/main/scala/org/scalajs/cli/Scalajsld.scala

```diff
@@ -138,58 +135,39 @@ object Scalajsld {
     }
 
     for (options <- parser.parse(args, Options())) {
-      val cpFiles = options.stdLib.toList ++ options.cp
-      // Load and resolve classpath
-      val cp = PartialClasspathBuilder.build(cpFiles).resolve()
-
-      // Write JS dependencies if requested
-      for (jsout <- options.jsoutput)
-        IO.concatFiles(WritableFileVirtualJSFile(jsout), cp.jsLibs.map(_.lib))
-
-      // Link Scala.js code
+      val classpath = options.stdLib.toList ++ options.cp
+      val irContainers = IRFileCache.IRContainer.fromClasspath(classpath)
+
+      // Warn if writing JS dependencies was requested.
+      if (options.jsoutput) {
+        Console.err.println(
+            "Support for the --jsoutput flag has been dropped. " +
+            "JS dependencies will not be written to disk. " +
+            "Comment on https://github.com/scala-js/scala-js/issues/2163 " +
+            "if you rely on this feature.")
+      }
+
+      val semantics =
+        if (options.fullOpt) options.semantics.optimized
+        else options.semantics
+
+      val frontendConfig = LinkerFrontend.Config()
+        .withBypassLinkingErrors(options.bypassLinkingErrors)
+        .withCheckIR(options.checkIR)
+
+      val backendConfig = LinkerBackend.Config()
+        .withRelativizeSourceMapBase(options.relativizeSourceMap)
+        .withPrettyPrint(options.prettyPrint)
+
+      val linker = Linker.linker(semantics, options.outputMode,
+          options.sourceMap, parallel = true, full = options.fullOpt,
+          frontendConfig, backendConfig)
+
+      val logger = new ScalaConsoleLogger(options.logLevel)
       val outFile = WritableFileVirtualJSFile(options.output)
-      if (options.fullOpt)
-        fullOpt(cp, outFile, options)
-      else
-        fastOpt(cp, outFile, options)
-    }
-  }
+      val cache = (new IRFileCache).newCache
```

Comment:
The cache is the only way we currently have to extract IR from `IRContainer`s. If you want, I can create a simpler extractor.